test/gtest/ucp/test_ucp_proto_select.cc

```diff
@@ -0,0 +1,386 @@
+        ucs_status_t status  = self->m_mock.actual_call(&tl->iface_open, md,
+                                                        worker, params, config,
+                                                        iface_p);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        uct_base_iface_t *base    = ucs_derived_of(*iface_p, uct_base_iface_t);
+        self->m_iface_names[base] = std::string(params->mode.device.tl_name) +
+                                    "/" + params->mode.device.dev_name;
+
+        self->m_mock.setup(&base->internal_ops->iface_estimate_perf,
+                           iface_estimate_perf_mock);
+        self->m_mock.setup(&(*iface_p)->ops.iface_query, iface_query_mock);
+        return status;
+    }
+
+    static ucs_status_t
+    iface_estimate_perf_mock(uct_iface_h iface, uct_perf_attr_t *perf_attr)
+    {
+        uct_base_iface_t *base = ucs_derived_of(iface, uct_base_iface_t);
+        mock_component *self   = mock_component_test::find(base->md->component);
+        ucs_status_t status    = self->m_mock.actual_call(
+                                    &base->internal_ops->iface_estimate_perf,
+                                    iface, perf_attr);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        auto it = self->m_perf_attrs_funcs.find(self->m_iface_names[base]);
+        if (it != self->m_perf_attrs_funcs.end()) {
+            (it->second)(*perf_attr);
+        }
+
+        return status;
+    }
+
+    static ucs_status_t
+    iface_query_mock(uct_iface_h iface, uct_iface_attr_t *iface_attr)
+    {
+        uct_base_iface_t *base = ucs_derived_of(iface, uct_base_iface_t);
+        mock_component *self   = mock_component_test::find(base->md->component);
+        ucs_status_t status    = self->m_mock.actual_call(
+                                    &iface->ops.iface_query, iface, iface_attr);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        auto it = self->m_iface_attrs_funcs.find(self->m_iface_names[base]);
+        if (it != self->m_iface_attrs_funcs.end()) {
+            (it->second)(*iface_attr);
+        }
+
+        return status;
+    }
+
+    uct_component_h                                     m_component;
+    ucs::mock                                          &m_mock;
+    std::unordered_map<uct_base_iface_t *, std::string> m_iface_names;
+    std::unordered_map<std::string, iface_attr_func_t>  m_iface_attrs_funcs;
+    std::unordered_map<std::string, perf_attr_func_t>   m_perf_attrs_funcs;
+};
+
+mock_component &mock_component_test::setup_mock(const std::string &name)
+{
+    uct_component_h component = find_uct_component(name);
+    if (nullptr == component) {
+        UCS_TEST_SKIP_R("Component " + name + " not found");
+    }
+
+    mock_component *mock_comp = find(component);
+    if (nullptr == mock_comp) {
+        mock_comp = new mock_component(component, m_mock);
```

Comment:
> Why can't each mock_component contain its own mock instance?

That would be a good idea, if we are 100% sure that component mocks are fully disjoint. But we cannot guarantee that, because different components may point to the same `uct_tl_t` instances, and therefore we may end mocking the same pointer multiple times.
What's wrong with mocking the same pointer twice?
It's a problem to revert mock changes to the original state, for example:
```C++
orig_fun_ptr = &orig_fun;
// remember CURRENT value (&orig_fun), restores on destroy: orig_fun_ptr = &orig_fun;
mock_func(&orig_fun_ptr, mock_fun);
// second mock, same ptr
// remember CURRENT value (&mock_fun), restores on destroy: orig_fun_ptr = &mock_fun;
mock_func(&orig_fun_ptr, mock_fun);
// which one is destroyed first?
```
Now you have a situation when behaviour depends on destructor order. We don't want that. We must use single mock instance for all UCT components

> Or static instance?

Yes, static would work.
However to me it seems that `m_mock` field in fixture is superior alternative, because it's automatically cleaned up upon test case completion. And also I try to avoid global variables when possible, IMO it's better to have explicit dependency injection here.